### ISSUE TITLE
I fine tuned my model through -ngl,  It doesn't seem to work

### ISSUE THREAD

**Issue by sohandsome1** [ggerganov/llama.cpp#5828]
./llama.cpp/build/bin/finetune --model-base ./llama/AI-ModelScope/chinese-alpaca-2-7b/ggml-model-f16.gguf  --lora-out ./lora-ggml-model-f16-LATEST.bin --train-data ./traindata.txt  --threads 8 -ngl 34  --adam-iter 156 --batch 8 --sample-start "\<s\>"

main: seed: 1709350890
main: model base = './llama/AI-ModelScope/chinese-alpaca-2-7b/ggml-model-f16.gguf'
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA A10, compute capability 8.6, VMM: yes
llama_model_loader: loaded meta data with 20 key-value pairs and 291 tensors from ./llama/AI-ModelScope/chinese-alpaca-2-7b/ggml-model-f16.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = LLaMA v2
llama_model_loader: - kv   2:                       llama.context_length u32              = 4096
llama_model_loader: - kv   3:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   4:                          llama.block_count u32              = 32
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 11008
llama_model_loader: - kv   6:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv   7:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv   8:              llama.attention.head_count_kv u32              = 32
llama_model_loader: - kv   9:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  10:                          general.file_type u32              = 1
llama_model_loader: - kv  11:                       tokenizer.ggml.model str              = llama
llama_model_loader: - kv  12:                      tokenizer.ggml.tokens arr[str,55296]   = ["\<unk\>", "\<s\>", "\</s\>", "\<0x00\>", "\<...
llama_model_loader: - kv  13:                      tokenizer.ggml.scores arr[f32,55296]   = [0.000000, 0.000000, 0.000000, 0.0000...
llama_model_loader: - kv  14:                  tokenizer.ggml.token_type arr[i32,55296]   = [2, 3, 3, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...
llama_model_loader: - kv  15:                tokenizer.ggml.bos_token_id u32              = 1
llama_model_loader: - kv  16:                tokenizer.ggml.eos_token_id u32              = 2
llama_model_loader: - kv  17:            tokenizer.ggml.padding_token_id u32              = 0
llama_model_loader: - kv  18:               tokenizer.ggml.add_bos_token bool             = true
llama_model_loader: - kv  19:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type  f16:  226 tensors
llm_load_vocab: mismatch in special tokens definition ( 889/55296 vs 259/55296 ).
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = llama
llm_load_print_meta: vocab type       = SPM
llm_load_print_meta: n_vocab          = 55296
llm_load_print_meta: n_merges         = 0
llm_load_print_meta: n_ctx_train      = 4096
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 32
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 1
llm_load_print_meta: n_embd_k_gqa     = 4096
llm_load_print_meta: n_embd_v_gqa     = 4096
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-05
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: n_ff             = 11008
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 10000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_yarn_orig_ctx  = 4096
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: model type       = 7B
llm_load_print_meta: model ftype      = F16
llm_load_print_meta: model params     = 6.93 B
llm_load_print_meta: model size       = 12.91 GiB (16.00 BPW) 
llm_load_print_meta: general.name     = LLaMA v2
llm_load_print_meta: BOS token        = 1 '\<s\>'
llm_load_print_meta: EOS token        = 2 '\</s\>'
llm_load_print_meta: UNK token        = 0 '\<unk\>'
llm_load_print_meta: PAD token        = 0 '\<unk\>'
llm_load_print_meta: LF token         = 13 '\<0x0A\>'
llm_load_tensors: ggml ctx size =    0.22 MiB
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 33/33 layers to GPU
llm_load_tensors:        CPU buffer size =   432.00 MiB
llm_load_tensors:      CUDA0 buffer size = 12785.02 MiB
................................................................................................
llama_new_context_with_model: n_ctx      = 512
llama_new_context_with_model: freq_base  = 10000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:      CUDA0 KV buffer size =   256.00 MiB
llama_new_context_with_model: KV self size  =  256.00 MiB, K (f16):  128.00 MiB, V (f16):  128.00 MiB
llama_new_context_with_model:  CUDA_Host input buffer size   =    10.01 MiB
llama_new_context_with_model:      CUDA0 compute buffer size =   116.00 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =     8.00 MiB
llama_new_context_with_model: graph splits (measure): 2
main: init model
print_params: n_vocab               : 55296
print_params: n_ctx                 : 128
print_params: n_embd                : 4096
print_params: n_ff                  : 11008
print_params: n_head                : 32
print_params: n_head_kv             : 32
print_params: n_layer               : 32
print_params: norm_rms_eps          : 0.000010
print_params: rope_freq_base        : 10000.000000
print_params: rope_freq_scale       : 1.000000
print_lora_params: n_rank_attention_norm : 1
print_lora_params: n_rank_wq             : 4
print_lora_params: n_rank_wk             : 4
print_lora_params: n_rank_wv             : 4
print_lora_params: n_rank_wo             : 4
print_lora_params: n_rank_ffn_norm       : 1
print_lora_params: n_rank_ffn_gate       : 4
print_lora_params: n_rank_ffn_down       : 4
print_lora_params: n_rank_ffn_up         : 4
print_lora_params: n_rank_tok_embeddings : 4
print_lora_params: n_rank_norm           : 1
print_lora_params: n_rank_output         : 4
main: total train_iterations 0
main: seen train_samples     0
main: seen train_tokens      0
main: completed train_epochs 0
main: lora_size = 86354720 bytes (82.4 MB)
main: opt_size  = 128829424 bytes (122.9 MB)
main: opt iter 0
main: input_size = 226496544 bytes (216.0 MB)
main: compute_size = 14940127840 bytes (14248.0 MB)
main: evaluation order = RIGHT_TO_LEFT
main: tokenize training data from ./traindata.txt
main: sample-start: \<s\>
main: include-sample-start: false
tokenize_file: warning: found 5 samples (max length 142) that exceed context length of 128. samples will be cut off.
tokenize_file: warning: found 618 samples (min length 68) that are shorter than context length of 128.
tokenize_file: total number of samples: 624
main: number of training tokens: 54703
main: number of unique tokens: 3074
main: train data seems to have changed. restarting shuffled epoch.
main: begin training
main: work_size = 1769984 bytes (1.7 MB)
train_opt_callback: iter=     0 sample=1/624 sched=0.000000 loss=0.000000 |>

**CPU usage**
![1709352463840](https://github.com/ggerganov/llama.cpp/assets/13551113/cacd1ed0-b38c-43d4-b33c-62fbbdccd2eb)

**GPU usage**
![1709352994972](https://github.com/ggerganov/llama.cpp/assets/13551113/eb899110-540f-422f-a43c-2d474c825dc7)

**Comment by Tachyon5** [comment]
The ggml-cuda backend doesn't support finetune. It errored for me when I ran a similar command to the one you ran.  Looks like deqantization doesn't happen in the cuda backend.

**Comment by teleprint-me** [comment]
I'm interested in this issue because there seems to be an emerging pattern Im unable to currently verify.

I noticed speedups if I compile woth GPU, and slowdowns when compiled for CPU, but theres no sign of GPU usage when running training. Same for finetune, and possibly no even flash attention (cant verify this because i do not have a supported card. Ill have to test this when i have some funds to get a runpod for a day or two).
